ast/src/analyzed/mod.rs

```diff
@@ -625,21 +628,51 @@ impl PublicDeclaration {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct SelectedExpressions<Expr> {
+    pub selector: Option<Expr>,
+    pub expressions: Vec<Expr>,
+}
+
+impl<Expr> Default for SelectedExpressions<Expr> {
+    fn default() -> Self {
+        Self {
+            selector: Default::default(),
+            expressions: vec![],
+        }
+    }
+}
+
+impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
+    /// Returns an iterator over all (top-level) expressions in this SelectedExpressions.
+    fn children(&self) -> Box<dyn Iterator<Item = &Expr> + '_> {
+        Box::new(self.selector.iter().chain(self.expressions.iter()))
+    }
+    /// Returns an iterator over all (top-level) expressions in this SelectedExpressions.
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut Expr> + '_> {
+        Box::new(self.selector.iter_mut().chain(self.expressions.iter_mut()))
+    }
+}
+
 #[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize, JsonSchema)]
-pub struct Identity<Expr> {
+pub struct Identity<SelectedExpressions> {
     /// The ID is globally unique among identities.
     pub id: u64,
     pub kind: IdentityKind,
     pub source: SourceRef,
     /// For a simple polynomial identity, the selector contains
     /// the actual expression (see expression_for_poly_id).
-    pub left: SelectedExpressions<Expr>,
-    pub right: SelectedExpressions<Expr>,
+    pub left: SelectedExpressions,
+    pub right: SelectedExpressions,
 }
 
-impl<Expr> Identity<Expr> {
+impl<T> Identity<SelectedExpressions<AlgebraicExpression<T>>> {
```

Comment:
Why not for generic `Expr`, i.e. `SelectedExpressions<Expr>`?

Reply:
The problem here is that now there are two SelectedExpression (`parsed` and `analyzed`) and in some cases functions specific to SelectedExpressions, such as `contains_next_ref`, are call, so you end up with a lot more duplicate code.

Although maybe a bit "uglier", I think this solution is cleaner.